File: src/DiagToolsValidatorFSharp/DiagToolsValidator.Core/Functionality/CommandLineTool.fs
namespace DiagToolsValidator.Core.CoreFunctionality

open System
open System.Diagnostics
open System.Text
open System.Collections.Generic
open System.Threading


module CommandLineTool =
    type CommandInvoker (environment: Dictionary<string, string>,
                         workDirectory: string,
                         fileName: string,
                         argument: string,
                         redirectStdOutErr: bool) =
        let _stdout = new StringBuilder()
        let _stderr = new StringBuilder()

        let proc = new Process()

        do proc.StartInfo.FileName <- fileName
        do proc.StartInfo.Arguments <- argument
        do proc.StartInfo.UseShellExecute <- false
        do proc.StartInfo.RedirectStandardInput <- true
        do proc.StartInfo.RedirectStandardOutput <- redirectStdOutErr
        do proc.StartInfo.RedirectStandardError <- redirectStdOutErr
        do proc.StartInfo.WorkingDirectory <- workDirectory

        do proc.StartInfo.EnvironmentVariables.Clear()
        do proc.StartInfo.EnvironmentVariables["Path"] <- Environment.GetEnvironmentVariable("Path")
        do for key in environment.Keys do
            proc.StartInfo.EnvironmentVariables[key] <- environment[key]

        let outputDataRecorder (sender: obj) (args: DataReceivedEventArgs) =
            let line = args.Data
            if not (isNull line) then
                _stdout.AppendLine $"   {line}" |> ignore
        
        let errorDataRecorder (sender: obj) (args: DataReceivedEventArgs) =
            let line = args.Data
            if not (isNull line) then
                _stderr.AppendLine $"   {line}" |> ignore

        do 
            if redirectStdOutErr
            then 
                proc.OutputDataReceived.AddHandler(outputDataRecorder)
                proc.ErrorDataReceived.AddHandler(errorDataRecorder)
            
        member val Command: string = $"{fileName} {argument}" with get
        member val StandardOutput: StringBuilder = _stdout with get
        member val StandardError: StringBuilder = _stderr with get
        member val Proc: Process = proc with get
        

    let PrinteOutputData (sender: obj) (args: DataReceivedEventArgs) =
        let line = $"   {args.Data}"
        if not (isNull line) then
            printfn "%s" line


    let IgnoreOutputData (sender: obj) (args: DataReceivedEventArgs) = ignore()
      
      
    let PrintErrorData (sender: obj) (args: DataReceivedEventArgs) =
        let line = $"   {args.Data}"
        if not (isNull line) then
            printfn "%s" line


    let IgnoreErrorData (sender: obj) (args: DataReceivedEventArgs) = ignore()        


    let RunCommand (fileName: string)
                   (argument: string)
                   (workDirectory: string)
                   (environment: Dictionary<string, string>)
                   (redirectStdOutErr: bool)
                   (outputDataReceivedHandler: obj -> DataReceivedEventArgs -> unit) 
                   (errorDataReceivedHandler: obj -> DataReceivedEventArgs -> unit) 
                   (waitForExit: bool)=
        try
            let commandInvoker = new CommandInvoker(environment,
                                                    workDirectory,
                                                    fileName,
                                                    argument,
                                                    redirectStdOutErr)
            printfn "Run command: %s" commandInvoker.Command
            if redirectStdOutErr then
                commandInvoker.Proc.OutputDataReceived.AddHandler outputDataReceivedHandler
                commandInvoker.Proc.ErrorDataReceived.AddHandler errorDataReceivedHandler

            commandInvoker.Proc.Start() |> ignore
            if redirectStdOutErr then
                commandInvoker.Proc.BeginOutputReadLine()
                commandInvoker.Proc.BeginErrorReadLine()

            if waitForExit
            then 
                commandInvoker.Proc.WaitForExit()

            Choice1Of2 commandInvoker
        with ex ->
            ex.Data.Add("RunCommand", $"Fail to run command: {fileName} {argument}")
            Choice2Of2 ex
    

    let TerminateCommandInvoker (commandInvoker: CommandInvoker) =
        // make sure the process is started
        try
            commandInvoker.Proc.Kill(true)
            commandInvoker.Proc.WaitForExit()
            while not (commandInvoker.Proc.HasExited) do
                Thread.Sleep(1000)
            Choice1Of2 commandInvoker
        with ex ->
            ex.Data.Add("TerminateCommandInvoker", $"Fail to terminate process: {commandInvoker.Command}")
            Choice2Of2 ex

File: src/DiagToolsValidationRunner/DiagToolsValidationRunner.Core/Functionality/CommandLineTool.cs
using System.Diagnostics;
using System.Text;

namespace DiagToolsValidationRunner.Core.Functionality
{
    public class CommandInvokeResult
    {
        public string Command { get; init; }
        public string StandardOutput { get; init; }
        public string StandardError { get; init; }
        public int ProcessID { get; init; }
        public Exception? Exn { get; init; }

        public CommandInvokeResult(string command, string stdout, string stderr, int pid, Exception? ex = null)
        {
            this.Command = command;
            this.StandardOutput = stdout;
            this.StandardError = stderr;
            this.ProcessID = pid;
            this.Exn = ex;
        }
    }

    public static class CommandInvokeTaskRunner
    {
        public static void Run(string loggerPath,
                               IEnumerable<CommandInvokeResult> commandInvokeTask,
                               bool ignoreError = false)
        {
            IEnumerator<CommandInvokeResult> enumrator = commandInvokeTask.GetEnumerator();
            while (true)
            {
                try
                {
                    if (!enumrator.MoveNext())
                    {
                        // Break when move to end
                        break;
                    }
                    CommandInvokeResult result = enumrator.Current;
                    StringBuilder logContent = new();
                    logContent.AppendLine($"Run Command: {result.Command}");
                    logContent.AppendLine(result.StandardOutput);
                    logContent.AppendLine(result.StandardError);
                    if (result.Exn != null)
                    {
                        logContent.AppendLine($"Error Message:{result.Exn.Message}");
                        logContent.AppendLine($"Stack Trace:\n{result.Exn.StackTrace}");
                        logContent.AppendLine($"Inner Exception:\n:{result.Exn.InnerException}");
                    }
                    logContent.AppendLine("\n");
                    File.AppendAllText(loggerPath, logContent.ToString());

                    if (!String.IsNullOrEmpty(result.StandardError) || result.Exn != null)
                    {
                        if (!ignoreError)
                        {
                            Console.WriteLine($"Run Command {result.Command} but get error! See {loggerPath} for details.");
                            break;
                        }
                    }
                }
                catch (Exception ex)
                {
                    StringBuilder logContent = new();
                    logContent.AppendLine($"Run into error: {ex.Message}");
                    logContent.AppendLine($"Stack Trace:\n{ex.StackTrace}");
                    logContent.AppendLine($"Inner Exception:\n{ex.InnerException}");
                    File.AppendAllText(loggerPath, logContent.ToString());
                    break;
                }
            }
        }

        public static void RecordSingle(string loggerPath, CommandInvokeResult result)
        {
            StringBuilder logContent = new();
            logContent.AppendLine($"Run Command: {result.Command}");
            logContent.AppendLine(result.StandardOutput);
            logContent.AppendLine(result.StandardError);
            if (result.Exn != null)
            {
                logContent.AppendLine($"Error Message:{result.Exn.Message}");
                logContent.AppendLine($"Stack Trace:\n{result.Exn.StackTrace}");
                logContent.AppendLine($"Inner Exception:\n:{result.Exn.InnerException}");
            }
            logContent.AppendLine("\n");
            File.AppendAllText(loggerPath, logContent.ToString());
        }
    }

    public class CommandInvoker: IDisposable
    {
        public static void PrintReceivedData(object sender, DataReceivedEventArgs args)
        {
            string? content = args.Data;
            if (!String.IsNullOrEmpty(content))
            {
                Console.WriteLine($"    {content}");
            }
        }
        
        private StringBuilder stdout;
        private StringBuilder stderr;
        private Process proc;

        public readonly string Command;

        public string StandardOutput
        {
            get { return stdout.ToString(); }
        }

        public string StandardError
        {
            get { return stderr.ToString(); }
        }

        public Process InvokedProcess
        {
            get { return proc; }
        }

        public CommandInvoker(string fileName,
                              string argument,
                              Dictionary<string, string> env,
                              string workDirectory = "")
        {
            // Initialize
            stdout = new();
            stderr = new();
            Command = $"{fileName} {argument}";
            proc = new();
            proc.StartInfo.FileName = fileName;
            proc.StartInfo.Arguments = argument;
            proc.StartInfo.UseShellExecute = false;
            proc.StartInfo.RedirectStandardInput = true;
            proc.StartInfo.WorkingDirectory = workDirectory;

            proc.StartInfo.EnvironmentVariables.Clear();
            proc.StartInfo.EnvironmentVariables["Path"] = Environment.GetEnvironmentVariable("Path");
            foreach (var key in env.Keys)
            {
                proc.StartInfo.EnvironmentVariables[key] = env[key];
            }
        }

        public void InvokeCommandWithOutWaitingForExit(bool redirectStdOutErr = true)
        {
            if (redirectStdOutErr)
            {
                proc.OutputDataReceived += (sender, args) =>
                {
                    string? line = args.Data;
                    if (!String.IsNullOrEmpty(line))
                    {
                        stdout.AppendLine(line);
                    }
                };
                proc.ErrorDataReceived += (sender, args) =>
                {
                    string? line = args.Data;
                    if (!String.IsNullOrEmpty(line))
                    {
                        stderr.AppendLine(line);
                    }
                };
            }

            proc.StartInfo.RedirectStandardOutput = redirectStdOutErr;
            proc.StartInfo.RedirectStandardError = redirectStdOutErr;

            Console.WriteLine($"\nRun command: {Command}");
            proc.Start();
            proc.BeginOutputReadLine();
            proc.BeginErrorReadLine();
        }

        public CommandInvokeResult InvokeCommand(bool redirectStdOutErr)
        {
            try
            {
                InvokeCommandWithOutWaitingForExit(redirectStdOutErr);

                int pid = proc.Id;
                proc.WaitForExit();
                proc.Dispose();
                return new(Command, stdout.ToString(), stderr.ToString(), pid);
            }
            catch (Exception ex)
            {
                return new(Command, stdout.ToString(), stderr.ToString(), -1, ex);
            }
        }

        public void Dispose()
        {
            proc.Kill(true);
            proc.WaitForExit();
            proc.Dispose();
        }
    }
}
